dev/bots/prepare_package.dart

```diff
@@ -0,0 +1,264 @@
+typedef ProcessResult ProcessRunner(
+  String executable,
+  List<String> arguments, {
+  String workingDirectory,
+  Map<String, String> environment,
+  bool includeParentEnvironment,
+  bool runInShell,
+  Encoding stdoutEncoding,
+  Encoding stderrEncoding,
+});
+
+/// Error class for when a process fails to run, so we can catch
+/// it and provide something more readable than a stack trace.
+class ProcessFailedException extends Error {
+  ProcessFailedException([this.message, this.exitCode]);
+
+  String message = '';
+  int exitCode = 0;
+
+  @override
+  String toString() => message;
+}
+
+/// Creates a pre-populated Flutter archive from a git repo.
+class ArchiveCreator {
+  /// [tempDir] is the directory to use for creating the archive.  Will place
+  /// several GiB of data there, so it should have available space.
+  /// [outputFile] is the name of the output archive. It should end in either
+  /// ".tar.bz2" or ".zip".
+  /// The runner argument is used to inject a mock of [Process.runSync] for
+  /// testing purposes.
+  ArchiveCreator(this.tempDir, this.outputFile, {ProcessRunner runner})
+      : assert(outputFile.path.toLowerCase().endsWith('.zip') ||
+            outputFile.path.toLowerCase().endsWith('.tar.bz2')),
+        flutterRoot = new Directory(path.join(tempDir.path, 'flutter')),
+        _runner = runner ?? Process.runSync {
+    flutter = path.join(
+      flutterRoot.absolute.path,
+      'bin',
+      Platform.isWindows ? 'flutter.bat' : 'flutter',
+    );
+    environment = new Map<String, String>.from(Platform.environment);
+    environment['PUB_CACHE'] = path.join(flutterRoot.absolute.path, '.pub-cache');
+  }
+
+  final Directory flutterRoot;
+  final Directory tempDir;
+  final File outputFile;
+  final ProcessRunner _runner;
+  String flutter;
+  final String git = Platform.isWindows ? 'git.bat' : 'git';
+  final String zip = Platform.isWindows ? 'zip.exe' : 'zip';
+  final String tar = Platform.isWindows ? 'tar.exe' : 'tar';
+  Map<String, String> environment;
+
+  /// Clone the Flutter repo and make sure that the git environment is sane
+  /// for when the user will unpack it.
+  void checkoutFlutter(String revision) {
+    // We want the user to start out the in the 'master' branch instead of a
+    // detached head. To do that, we need to make sure master points at the
+    // desired revision.
+    runGit(<String>['clone', '-b', 'master', CHROMIUM_REPO], workingDirectory: tempDir);
+    runGit(<String>['reset', '--hard', revision]);
+
+    // Make the origin point to github instead of the chromium mirror.
+    runGit(<String>['remote', 'remove', 'origin']);
+    runGit(<String>['remote', 'add', 'origin', GITHUB_REPO]);
+  }
+
+  /// Prepare the archive repo so that it has all of the caches warmed up and
+  /// is configured for the user to being working.
+  void prepareArchive() {
+    runFlutter(<String>['update-packages']);
+    runFlutter(<String>['doctor']);
```

Comment:
i'd put the doctor call first if you're going to have a doctor call.

`flutter doctor` has no side-effects (other than the same side-effects that launching any `flutter` command has, e.g. compiling the flutter tool).